### ISSUE TITLE
Does optimization/acquisition function operate in search space or embedding space?

### ISSUE THREAD

**Issue by brandon-holt** [emdgroup/baybe#307]
To help me understand this I'll describe two scenarios and you tell me what the mathematical differences will be. This is in the context of discrete search space.

1. We have 10 features that we use to construct our search space with 10 discrete paramaters.
2. Using the same data set, we construct our search space with 1 `CustomDiscreteParameter` and define the encoding as a vector of length 10, which uses the features from scenario 1. 

In the specific context of a simulate scenarios/experiment exercise, where you only allow recommendations that already exist in your dataset, wouldn't scenario 2 be the most compact way to construct the search space? 

Further, are there any differences between the two scenarios in how the acquisition function will recommend new experiments?

However, I see that scenario 2 wouldnt work for the use case where you want to generate novel combinations of features.

**Comment by AdrianSosic** [comment]
Hi @brandon-holt 👋🏼 

**TL;DR: No, there are no differences. Your two representations are exactly equivalent from a modeling perspective, but I'd argue that the latter has some practical disadvantages.**

The fact that both representations are possible lies in the nature of discrete spaces and that they are always *countable* (no matter if the space is finite or infinite – after all, that's what makes a discrete space *discrete*). Basically, it means that you can always flatten an arbitrarily structured discrete space into a one that can enumerate using a one-dimensional index. You might have seen this people using this argument in the context of measuring the cardinality of sets:
https://en.wikipedia.org/wiki/Pairing_function

It effectively boils down to "re-labeling" your points. Here a simple example for a "two-dimensional" discrete space. Notice the quotation marks, since I can just as well compress it into a "one-dimensional" space by throwing away the axes, which is exactly the point of this example:
<img width="815" alt="image" src="https://github.com/emdgroup/baybe/assets/23265127/18bc8ede-6740-4675-8cb8-e15435a99263">

In BayBE, you can generate these two representations as follows
```python
from baybe.parameters import NumericalDiscreteParameter
from baybe.parameters.custom import CustomDiscreteParameter
from baybe.searchspace.core import SearchSpace

parameters = [
    NumericalDiscreteParameter("x", [1, 2]),
    NumericalDiscreteParameter("y", [3, 4]),
]
s1 = SearchSpace.from_product(parameters)

data = s1.discrete.exp_rep
data.index = map(str, data.index)
c = CustomDiscreteParameter("cantor", data)
s2 = SearchSpace.from_product([c])

print(s1)
print(s2)
```
If you look at the printed output, you'll notice that the `computational representation` of both search spaces is identical (apart from the column labels), and this makes the two spaced behave identically when used with our models – it is the `computational representation` that enters the created surrogate models / acquisition function. So as long as you ensure they don't differ, you'll get the same results. Notice though, that in other situations (e.g. where you use a `CategoricalParameter`, the obtained `computational representation` of that parameter will depend on the chosen encoding).

But while they are mathematically equivalent, the latter representation comes with the disadvantage that you lose the inherent structure of the space. Having access it is useful, for instance, when restricting the space using constraints, or when analyzing results (in general: whenever you want to perform operations on that structure). 

#### Example
Consider the case where you want to filter all points containing `x_1` from your search space. You can then simply create a corresponding constraint (such as a `DiscreteExcludeConstraint`) that operates on the named axis, to remove the undesired elements. If you want to achieve the same using the "flattened" representation, the only way to do this is really to explicitly enumerate all elements you want to remove one by one, since the structure information is no longer available. The same problem occurs, for example, when you want to analyze feature importance of your learned model.

Does this make sense to you?

**Comment by brandon-holt** [comment]
Yes this totally makes sense!! But it would in fact be true that the latter representation would have a smaller memory footprint?

**Comment by AdrianSosic** [comment]
No, not really. Or where exactly do you see the potential reduction?

If you do it right, both representations hold the exact same number of choices, both in experimental and in computational representation. Perhaps your confusion lies in the assumption of having to go via `from_product` only to narrow down your search space afterward (which would indeed have an unnecessary large *temporary* footprint during the intermediate phase before the filtering but the final result would still be the same). But remember: if you don't need a product search space, simply don't create one in the first place. There are many other options how you can create a space such that you get the desired representations directly, without having to go via a two-step approach. A one-to-one replacement would be the `from_dataframe` constructor in your case, which you can directly pass the dataframe that you would otherwise pass to the `CustomDiscreteParameter` instead.

**Comment by brandon-holt** [comment]
@AdrianSosic Okay, got it, thank you for the clarification!
